File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/support/ExtJpaSupport.java
package top.bettercode.summer.data.jpa.support;

/**
 * @author Peter Wu
 */
public interface ExtJpaSupport {

  void setSoftDeleted(Object entity);

  void setUnSoftDeleted(Object entity);

  boolean supportSoftDeleted();

  boolean isSoftDeleted(Object entity);

  boolean softDeletedSeted(Object entity);

  Class<?> getSoftDeletedPropertyType();

  String getSoftDeletedPropertyName();

  String getLastModifiedDatePropertyName();

  String getVersionPropertyName();

  Object getLastModifiedDateNowValue();

  Object getVersionIncValue();

  Object getSoftDeletedTrueValue();

  Object getSoftDeletedFalseValue();

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/support/ExtJpaSupport.kt
package top.bettercode.summer.data.jpa.support

import jakarta.persistence.metamodel.SingularAttribute
import top.bettercode.summer.data.jpa.metamodel.LastModifiedByAttribute
import top.bettercode.summer.data.jpa.metamodel.LastModifiedDateAttribute
import top.bettercode.summer.data.jpa.metamodel.LogicalDeletedAttribute
import top.bettercode.summer.data.jpa.metamodel.VersionAttribute

/**
 * @author Peter Wu
 */
interface ExtJpaSupport<T> {

    val idAttribute: SingularAttribute<T, *>?

    val logicalDeletedSupported: Boolean

    val logicalDeletedAttribute: LogicalDeletedAttribute<T, *>?

    val lastModifiedDateAttribute: LastModifiedDateAttribute<T, *>?

    val lastModifiedByAttribute: LastModifiedByAttribute<T, *>?

    val versionAttribute: VersionAttribute<T, *>?
}
